File: 3. Prolog/Task-3.pl
% Task 4, Section 3:
% For a given word w, determine if the finite automaton accepts
% at least one word of the form x++w (for some word x, possibly empty).
% If yes, provide an example of such word xw.

% --- Automaton: DFA over {'0','1'} accepting binary numbers divisible by 3 ---
% State 0: remainder 0 (initial, accepting)
% State 1: remainder 1
% State 2: remainder 2

initial(0).
final(0).
alphabet(['0', '1']).

trans(0, '0', 0).
trans(0, '1', 1).
trans(1, '0', 2).
trans(1, '1', 0).
trans(2, '0', 1).
trans(2, '1', 2).

% --- run_word(+State, +Word, -ReachedState) ---
% Simulates the automaton reading Word starting from State.
run_word(S, [], S).
run_word(S, [H|T], F) :-
    trans(S, H, S1),
    run_word(S1, T, F).

% --- w_good(+W, ?Q) ---
% State Q is "w-good" if reading W from Q leads to an accepting state.
w_good(W, Q) :-
    run_word(Q, W, F),
    final(F).

% --- all_states(-States) ---
% Collect all states mentioned in transitions.
all_states(States) :-
    findall(Q, (trans(Q, _, _) ; trans(_, _, Q)), All),
    list_to_set(All, States).

% --- bfs(+Queue, +Visited, +Targets, -RevPrefix) ---
% BFS from initial state toward any state in Targets.
% Queue: list of (State, ReversedPath) pairs.
bfs([(S, RevPath)|_], _, Targets, RevPath) :-
    member(S, Targets), !.
bfs([(S, _)|Rest], Visited, Targets, RevPath) :-
    member(S, Visited), !,
    bfs(Rest, Visited, Targets, RevPath).
bfs([(S, RevPath)|Rest], Visited, Targets, Result) :-
    alphabet(Alph),
    findall(
        (S1, [C|RevPath]),
        (   member(C, Alph),
            trans(S, C, S1),
            \+ member(S1, [S|Visited])
        ),
        Nexts
    ),
    append(Rest, Nexts, NewQueue),
    bfs(NewQueue, [S|Visited], Targets, Result).

% --- can_extend_left(+W, -XW) ---
% Main predicate: determines if the automaton accepts some word XW = X ++ W.
% Returns XW (as a list of characters) if yes, fails otherwise.
can_extend_left(W, XW) :-
    all_states(AllStates),
    findall(Q, (member(Q, AllStates), w_good(W, Q)), Targets),
    Targets \= [],
    initial(Init),
    bfs([(Init, [])], [], Targets, RevPrefix),
    reverse(RevPrefix, Prefix),
    append(Prefix, W, XW).

File: 3. Haskell/Task-3.hs
-- Task 5, Section 3:
-- For a given word w, determine if the finite automaton accepts
-- at least one word of the form w++y (for some word y, possibly empty).
-- If yes, provide an example of such word w++y.

module Task3 where

import Data.List (nub)

type State  = Int
type Symbol = Char

-- | A nondeterministic finite automaton (NFA / DFA)
data FA = FA
  { delta  :: [(State, Symbol, State)]  -- transition relation
  , q0     :: State                     -- initial state
  , qFinal :: [State]                   -- set of accepting states
  , sigma  :: [Symbol]                  -- alphabet
  } deriving (Show)

-- | All states reachable from state q by reading one symbol c
step :: FA -> State -> Symbol -> [State]
step fa q c = [r | (s, a, r) <- delta fa, s == q, a == c]
  
-- | Set of states reached after reading word w from the initial state
runWord :: FA -> [Symbol] -> [State]
runWord fa = foldl nextLayer [q0 fa]
  where
    nextLayer qs c = nub $ concatMap (\q -> step fa q c) qs

-- | BFS from a set of start states.
--   Finds the shortest suffix y such that reading y from some start state
--   leads to an accepting state.
--   Returns Just y if such y exists, Nothing otherwise.
findSuffix :: FA -> [State] -> Maybe [Symbol]
findSuffix fa starts = bfs [(q, []) | q <- nub starts] []
  where
    bfs [] _ = Nothing
    bfs queue visited =
      case filter (\(q, _) -> q `elem` qFinal fa) queue of
        (_, y) : _ -> Just y      -- found: return the suffix
        []         ->
          let visited'  = visited ++ map fst queue
              nextQueue = nubFirst
                [ (r, path ++ [c])
                | (q, path) <- queue  -- for each current state
                , c         <- sigma fa  -- try every symbol
                , r         <- step fa q c  -- compute next state
                , r `notElem` visited'  -- skip already-visited states
                ]
          in bfs nextQueue visited'

    -- keep only the first (shortest-path) occurrence of each state
    nubFirst = foldr
      (\x acc -> if fst x `elem` map fst acc then acc else x : acc) []

-- | For a given automaton fa and word w,
--   determine whether fa accepts some word of the form w ++ y.
--   Returns Just (w ++ y) if yes, Nothing otherwise.
canExtend :: FA -> [Symbol] -> Maybe String
canExtend fa w = fmap (w ++) (findSuffix fa (runWord fa w))
